### PR TITLE
Issue 816: Auto-detect CSV Dates

### PR DESCRIPTION
**Pull request by hawkfish**
As discussed in Issue 816, add code to try to detect some common numeric date formats when sniffing a CSV file.

### CODE REVIEW

File: src/include/duckdb/execution/operator/persistent/buffered_csv_reader.hpp [hunk 1]
```diff
@@ -84,13 +85,9 @@ struct BufferedCSVReaderOptions {
...
 	idx_t num_samples = 10;
 	//! The date format to use (if any is specified)
-	StrpTimeFormat date_format;
-	//! Whether or not a date format is specified
-	bool has_date_format = false;
-	//! The timestamp format to use (if any is specified)
-	StrpTimeFormat timestamp_format;
-	//! Whether or not a timestamp format is specified
-	bool has_timestamp_format = false;
+	std::map<LogicalTypeId, StrpTimeFormat> date_format = {{LogicalTypeId::DATE, {}}, {LogicalTypeId::TIMESTAMP, {}}};
```
Comment (Mytherin) [inline src/include/duckdb/execution/operator/persistent/buffered_csv_reader.hpp:88]: Any specific reason to use map over unordered_map?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! It looks very good. I would like to see more tests with CSVs containing each of the different date formats if possible. Also some tests with CSVs with conflicting date formats (e.g. containing only 01-01-2020). What happens in the case where there are multiple potential formats?

Otherwise I think it looks great. Thanks again :)

**Comment by hawkfish** [comment]
I can add some more tests, no problem. Also, we can add `TIME` formats if you like as it is all table driven, but I don't think there's a lot to gain there.

The ambiguous format question is hard to test for because the output is the same either way. The code I wrote has prioritised formats, but the example you give would parse correctly for either `%m-%d-%Y` or` %d-%m-%Y`. Is there a way to dump the format used by `COPY` to the output?

File: src/include/duckdb/execution/operator/persistent/buffered_csv_reader.hpp (hunk 1 above)
Comment (hawkfish) [reply src/include/duckdb/execution/operator/persistent/buffered_csv_reader.hpp:88]: Yeah, I couldn't see any `hash` definition for `LogicalTypeId` (which is really uint8_t`?). I can poke at that some more.( I first considered using just a `vector` with empty slots, but the `enum` is not dense.`)

**Comment by Mytherin** [comment]
> I can add some more tests, no problem. Also, we can add `TIME` formats if you like as it is all table driven, but I don't think there's a lot to gain there.
> 
> The ambiguous format question is hard to test for because the output is the same either way. The code I wrote has prioritised formats, but the example you give would parse correctly for either `%m-%d-%Y` or` %d-%m-%Y`. Is there a way to dump the format used by `COPY` to the output?

Indeed 01-01-2020 would parse the same, but 01-02-2020 then :) I think rather than prioritised formats it might be better to throw an exception on an ambiguous date format being encountered, or falling back to parsing as a string so the user can later correct the parsing manually. 

There is no way to output the format used by the copy command at this point, although that might be useful for the auto detection indeed. Perhaps something along those lines could be added.

File: src/include/duckdb/execution/operator/persistent/buffered_csv_reader.hpp (hunk 1 above)
Comment (Mytherin) [reply src/include/duckdb/execution/operator/persistent/buffered_csv_reader.hpp:88]: I think the map is fine, just wondering about the reason :) In general it should only contain two keys anyway.

File: src/execution/operator/persistent/buffered_csv_reader.cpp [hunk 2]
```diff
@@ -455,6 +550,61 @@ vector<LogicalType> BufferedCSVReader::SniffCSV(vector<LogicalType> requested_ty
...
+									type_format_candidates.emplace_back(format_string);
+							}
+
+							//	initialise the first candidate
+							options.has_format[sql_type.id()] = true;
+							//	all formats are constructed to be valid
+							SetDateFormat(type_format_candidates.back(), sql_type.id());
+						}
+						// check all formats and keep the first one that works
+						int32_t date_parts[6];
```
Comment (Mytherin) [inline src/execution/operator/persistent/buffered_csv_reader.cpp:576]: This should be [7], not [6] :)

**Comment by Mytherin** [comment]
Thanks for the extra tests! I investigated the 32-bit Windows crash, and it seems to be an off-by-one in the date_parts array. As to why that bug only triggers on Windows 32-bit, I have no clue. Could you fix that? Then after the tests pass I think we can merge it.

**Comment by hawkfish** [comment]
Another victory for strong typing... We should change the argument to `Parse` to be a `struct` that has all the output variables. But that is a problem for another day.

File: src/execution/operator/persistent/buffered_csv_reader.cpp (hunk 2 above)
Comment (hawkfish) [reply src/execution/operator/persistent/buffered_csv_reader.cpp:576]: Another victory for strong typing... We should change the argument to Parse to be a struct that has all the output variables. But that is a problem for another day.

Comment (Mytherin) [reply src/execution/operator/persistent/buffered_csv_reader.cpp]: Indeed, good point.

**Pull request merged**